File: cpp/medium/Subsets.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {

  vector<vector<int>> ans;

public:
  vector<vector<int>> subsets(vector<int> &nums) {
    vector<int> subset;
    helper(nums, subset, 0);
    return ans;
  }

  void helper(vector<int> &nums, vector<int> &subset, int i) {
    if (i == nums.size()) {
      ans.push_back(subset);
      return;
    }
    helper(nums, subset, i + 1);
    subset.push_back(nums[i]);
    helper(nums, subset, i + 1);
    subset.pop_back();
  }
};

File: stonycodes70proplems/medium/Subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def backtrack(sub: List[int], idx: int):
            if idx == len(nums):
                ans.append(sub.copy())
                return

            backtrack(sub, idx + 1)
            sub.append(nums[idx])
            backtrack(sub, idx + 1)
            sub.pop()

        backtrack([], 0)

        return ans
